accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,119 @@
+
+This requires a mapping between the MSBuild property name, the full name of the feature switch for runtime configuration (which would show up in `.runtimeconfig.json`) and the read-only static property in the managed code which is used to branch the behavior. 
+
+*Note that it seems likely that some of switches may require 1:many mapping because they're targeting existing code which uses multiple properties to determine the presence of the feature.*
+
+Ultimately some of these switches may also need VS UI integration, but for now most of the feature switches should be perfectly fine with only MSBuild properties.
+
+The ability to specify these from MSBuild projects as properties also means that different templates and SDKs can set different defaults - so for example Blazor or Xamarin SDKs may choose to turn off some of these features by default.
+
+### Generate the right input for the linker in SDK
+Currently there's no integration of the substitutions feature in the linker with the `ILLink` task/targets. Introduce a standard way to  define the content of `substitutions.xml` from the SDK (MSBuild items/properties) and then implement the file generation in the `ILLink` task and pass it to the linker.
+
+*Or change linker to not use the XML format and instead take the input on command line.*
+
+
+## Open questions
+
+### Boolean only or more complex values
+Should the feature switches always represent simple on/off statements and thus be represented as boolean properties? In some cases the simple boolean may necessitate presence of multiple similar feature switches. For example the security algorithms case described above. In the extreme each such algorithm would have its own feature switch - this would obviously create confusion. We could potentially allow feature switches to have string values from a predefined set to solve this problem.
+
+### Which components own the mapping between different representations of feature switches
+There are 3 different representations to each feature switch
+* The MSBuild property - for example `InvariantGlobalization` - this is used by developers to turn the feature on/off.
+* The runtime property in `.runtimeconfig.json` - for example `System.Globalization.Invariant` - this is used at runtime to alter the behavior of the code (via the IL property below)
+* The IL property which is used in the code - for example `System.Globalization.GlobalizationMode.Invariant, System.Private.CoreLib` - this is used by the linker to fix the value of the property and propagate that to remove unused code.
+
+The mapping between MSBuild property and runtime property pretty much has to be part of the SDK (in the MSBuild somewhere) as the SDK already generates the `.runtimeconfig.json` and this needs to be part of it. The second mapping to the IL property can take several options:
```

Comment:
Would it make sense to get rid of this mapping and switch to unified system where the name is same everywhere? What is the naming pattern we would unify on if we were to do this?

The existing cases can be handled as hard-coded one off, or we can take a breaking change for them in .NET 5.

Reply:
It would be hard to unify on the name of the MSBuild property - and I don't think it's the right thing to do anyway. But we could unify the runtime and link-time names. The runtime name can be arbitrary as long as it's unique because we will need code to call the `TryGetSwitch` anyway. The link-time name needs to be able to tell linker what to do, so ideally the full name of the property.

So in the example of the globalization invariant the unified name would be `System.Globalization.GlobalizationMode.Invariant`. Such a string would work for the runtime property as well (it would not work for MSBuild since it's not very user friendly and it's not a valid MSBuild property anyway).

There are some draw backs though:
* Inability to have a 1:many mapping between the runtime property and the IL properties - makes it even more important to make the IL property with the right visibility.
* Inability to specify the assembly for the IL property making it a bit ambiguous. That said this could be used to "hack around" the above limitation of missing 1:many mapping (more than one assembly could define the property with the same namespace/type/name)

Reply:
You can use MSBuild items.